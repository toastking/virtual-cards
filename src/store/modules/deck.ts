import { Module } from 'vuex';
import { State } from '@/store';
import { firestoreAction } from 'vuexfire';
import { db } from '@/db';

export interface Deck {
  drawnCards: string[];
  currentCard?: string;
  id?: string; // Id from firestore
}

export interface DeckState {
  decks: Deck[];
  currentDeckIndex: number;
}

const baseDeck = createInitialDeck();

export const DeckModule: Module<DeckState, State> = {
  state: () => ({ decks: [], currentDeckIndex: 0 }),
  actions: {
    /** Draw a random card from a deck in the game */
    drawCard({ rootState, getters }) {
      // Get the drawn cards for the current deck
      const availableCards: string[] = getters.cardsStillLeft;
      const cardIndex = Math.floor(Math.random() * availableCards.length);
      const drawnCard = availableCards[cardIndex];
      const currentDeck: Deck = getters.currentDeck;
      const updatedDrawnCards = currentDeck.drawnCards.concat(drawnCard);

      // update the Deck value in firestore
      const { gameId } = rootState.game;
      if (gameId) {
        const fieldsToUpdate: Partial<Deck> = {
          currentCard: drawnCard,
          drawnCards: updatedDrawnCards,
        };
        return db
          .collection('games')
          .doc(gameId)
          .collection('decks')
          .doc(currentDeck.id)
          .update(fieldsToUpdate);
      }
    },
    /** Add a deck to firebase */
    addDeck({ rootState }) {
      const { gameId } = rootState.game;
      if (gameId) {
        const newDeck: Deck = { drawnCards: [] };
        return db
          .collection('games')
          .doc(gameId)
          .collection('decks')
          .add(newDeck);
      }
    },
    setupDeckBinding: firestoreAction(({ rootState, bindFirestoreRef }) => {
      const { gameId } = rootState.game;
      if (gameId) {
        return bindFirestoreRef(
          'decks',
          db
            .collection('games')
            .doc(gameId)
            .collection('decks')
        );
      }
    }),
  },
  getters: {
    /** Calculate the cards still left to draw in a given deck */
    cardsStillLeft(state): string[] {
      const drawnCards = new Set(
        state.decks[state.currentDeckIndex].drawnCards
      );
      return baseDeck.filter(card => !drawnCards.has(card));
    },
    /** Get the current deck we're working on */
    currentDeck(state): Deck {
      return state.decks[state.currentDeckIndex];
    },
  },
};

/** Make the initial deck value */
function createInitialDeck(): string[] {
  const ranks = [
    'A',
    '2',
    '3',
    '4',
    '5',
    '6',
    '7',
    '8',
    '9',
    'T',
    'J',
    'Q',
    'K',
  ];
  const suits = ['c', 'd', 'h', 's'];
  const deck: string[] = [];

  for (const suit of suits) {
    for (const rank of ranks) {
      deck.push(`${rank}${suit}`);
    }
  }
  return deck;
}
