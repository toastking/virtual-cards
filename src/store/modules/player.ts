import { Module } from 'vuex';
import { db } from '@/db';
import { State } from '@/store';
import { firestoreAction } from 'vuexfire';

export interface Player {
  name: string;
  id?: string;
}

export interface PlayerState {
  players: [];
}

/** Module to handle player state, like who's in the game and who is the current player. */
export const PlayerModule: Module<PlayerState, State> = {
  state: () => ({ currentPlayerId: null, players: [] }),
  actions: {
    /** Add a player to the firebase store */
    async addPlayer({ rootState }, player: Player) {
      const { gameId } = rootState.game;
      if (gameId) {
        return db
          .collection('games')
          .doc(gameId)
          .collection('players')
          .add(player);
      }
    },
    /** Sets up the firebase binding for the players list  */
    setupPlayerBinding: firestoreAction(
      async ({ bindFirestoreRef, rootState }) => {
        const { gameId } = rootState.game;
        if (gameId) {
          return bindFirestoreRef(
            'players',
            db
              .collection('games')
              .doc(gameId)
              .collection('players')
          );
        }
      }
    ),
  },
};
