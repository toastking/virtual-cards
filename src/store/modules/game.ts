import { db } from '@/db';
import router from '@/router';
import { State } from '@/store';
import { Module } from 'vuex';
import { firestoreAction } from 'vuexfire';
import { Player } from './player';

export enum LoadingStatus {
  NOT_STARTED,
  LOADING,
  OK,
  ERROR,
}

/** Represents the general state of a card game */
export interface Game {
  currentPlayer: string | null;
  gameCompleted: boolean;
  gameStarted: boolean;
}

export interface GameState {
  gameId: string | null;
  gameLoadingStatus: LoadingStatus;
  game: Game;
  // Loading for the request for the turn
  turnLoadingState: LoadingStatus;
}

/** Module to handle game state. This is things like the game ID and the completion state. */
export const GameModule: Module<GameState, State> = {
  state: () => ({
    gameId: null,
    gameLoadingStatus: LoadingStatus.NOT_STARTED,
    turnLoadingState: LoadingStatus.OK,
    game: {
      currentPlayer: null,
      gameCompleted: false,
      gameStarted: false,
    },
  }),
  mutations: {
    createGameSuccess(state, payload: { newGameId: string }) {
      state.gameId = payload.newGameId;
      state.gameLoadingStatus = LoadingStatus.OK;
    },
    turnRequestStarted(state) {
      state.turnLoadingState = LoadingStatus.LOADING;
    },
    turnRequestDone(state) {
      state.turnLoadingState = LoadingStatus.OK;
    },
  },
  actions: {
    /** Create a new game object and route to the lobby */
    async createGame({ dispatch }, payload: { hostPlayerName: string }) {
      const dummyGame: Game = {
        currentPlayer: null,
        gameCompleted: false,
        gameStarted: false,
      };
      const newGame = await db.collection('games').add(dummyGame);

      dispatch('joinGame', {
        gameId: newGame.id,
        playerName: payload.hostPlayerName,
      });
    },
    /** Joins a game by setting the gameId and creating a player with the player info */
    joinGame(
      { dispatch, commit },
      payload: { gameId: string; playerName: string }
    ) {
      commit('createGameSuccess', { newGameId: payload.gameId });

      const player: Player = { name: payload.playerName };
      dispatch('addUserPlayer', player);

      dispatch('routeToLobby', { newGameId: payload.gameId });
    },
    /** Changes the gameStarted boolean so the game is started  */
    startGame({ state, dispatch, rootState }) {
      const { gameId } = state;

      dispatch('addDeck');
      if (gameId) {
        // set game started to true and set the current player id to the first player
        const firstPlayerId = rootState.player.players[0].id;
        db.collection('games')
          .doc(gameId)
          .update({ gameStarted: true, currentPlayer: firstPlayerId });
      }
    },
    /** Setup the firebase binding for the game */
    setupGameBinding: firestoreAction(async ({ bindFirestoreRef, state }) => {
      const { gameId } = state;
      if (gameId) {
        return bindFirestoreRef('game', db.collection('games').doc(gameId), {
          reset: false,
        });
      }
    }),
    routeToLobby({}, payload: { newGameId: string }) {
      router.push(`lobby/${payload.newGameId}`);
    },
    routeToGame({ state }) {
      const { gameId } = state;
      if (gameId) {
        router.push({ name: 'game', params: { gameId } });
      }
    },
  },
};
