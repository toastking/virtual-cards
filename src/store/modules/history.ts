import { db } from '@/db';
import { State } from '@/store';
import { Module } from 'vuex';
import { firestoreAction } from 'vuexfire';
import { Player } from './player';
import { firestore } from 'firebase';

/** A record of what card was drawn by what user */
export interface HistoryEntry {
  playerId: string;
  /** String representation of a card based on vue-plaing-card */
  card: string;
  timestamp: firestore.Timestamp;
}
/** State for which card was drawn and when */
export interface HistoryState {
  history: HistoryEntry[];
}

export const HistoryModule: Module<HistoryState, State> = {
  actions: {
    /** Adds the history to a game */
    addHistory({ rootState }, payload: { player: Player; card: string }) {
      const historyEntry: HistoryEntry = {
        playerId: payload.player?.id ?? '',
        card: payload.card,
        timestamp: firestore.Timestamp.fromDate(new Date()),
      };

      const { gameId } = rootState.game;
      if (gameId) {
        return db
          .collection('games')
          .doc(gameId)
          .collection('history')
          .add(historyEntry);
      }
    },
    setupHistoryBinding: firestoreAction(({ rootState, bindFirestoreRef }) => {
      const { gameId } = rootState.game;
      if (gameId) {
        return bindFirestoreRef(
          'history',
          db
            .collection('games')
            .doc(gameId)
            .collection('history')
            .orderBy('timestamp')
        );
      }
    }),
  },
};
