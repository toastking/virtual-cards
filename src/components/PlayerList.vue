<template>
  <div class="box has-background-white-ter">
    <h2 class="title is-2">Players</h2>
    <ol class="list">
      <li class="player-list-item list-item" v-for="player in players" :key="player.id">
        <b-tooltip class="columns is-vcentered player-info" :label="player.name" animated>
          <div class="column is-narrow">
            <avatar :avatar="player.avatar" />
          </div>
          <div class="column player-name has-text-left is-size-4">
            <p>{{ player.name }}</p>
          </div>
          <div class="column is-narrow">
            <span v-if="isTurn(player)" class="turn-tag tag is-success is-medium">Turn</span>
          </div>
        </b-tooltip>
      </li>
    </ol>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import Avatar from '@/components/Avatar.vue';
import { mapState, mapGetters } from 'vuex';
import { State } from '@/store';
import { GameModule } from '../store/modules/game';
import { PlayerState, Player } from '@/store/modules/player';

export default Vue.extend({
  computed: mapState({
    players(state: State) {
      return state.player.players;
    },
  }),
  methods: {
    isTurn(player: Player) {
      return player.id === this.$store.state.game.game.currentPlayer;
    },
    ...mapGetters(['isYourTurn']),
  },
  components: { avatar: Avatar },
});
</script>

<style scoped>
.player-name {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.player-list-item {
  cursor: pointer;
}
</style>
