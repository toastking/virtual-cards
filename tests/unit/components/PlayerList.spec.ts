import PlayerList from '@/components/PlayerList.vue';
import { Game, GameState } from '@/store/modules/game';
import { Player, PlayerState } from '@/store/modules/player';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Buefy from 'buefy';
import { Store } from 'vuex-mock-store';

describe('PlayerList', () => {
  const player1: Player = { name: 'foo', id: 'currentId' };
  const player2: Player = { name: 'bar' };
  const playerState: Partial<PlayerState> = {
    players: [player1, player2],
    userPlayerId: 'currentId',
  };
  const gameDoc: Game = {
    gameCompleted: false,
    currentPlayer: '',
    gameStarted: true,
  };
  const game: Partial<GameState> = { game: gameDoc };
  const store = new Store({ state: { player: playerState, game } });
  // add other mocks here so they are accessible in every component
  const mocks = {
    $store: store,
  };

  const localVue = createLocalVue();
  localVue.use(Buefy);

  afterEach(() => {
    store.reset();
  });

  test('renders the list of players from the store', () => {
    const wrapper = shallowMount(PlayerList, { mocks, localVue });
    const playerNames = wrapper.findAll('.player-name');

    expect(playerNames.at(0).text()).toBe('foo');
    expect(playerNames.at(1).text()).toBe('bar');
  });

  test('shows a tag for the current player turn', () => {
    store.state.game.game!.currentPlayer = 'currentId';
    const wrapper = shallowMount(PlayerList, { mocks, localVue });
    const firstPlayer = wrapper.findAll('.player-info').at(0);
    const tag = firstPlayer.find('.tag');

    expect(tag.exists()).toBe(true);
  });
});
