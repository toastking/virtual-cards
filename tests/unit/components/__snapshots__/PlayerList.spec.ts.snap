// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerList renders the list of players from the store 1`] = `
<div
  class="box has-background-white-ter"
>
  <h2
    class="title is-2"
  >
    Players
  </h2>
   
  <ol
    class="list"
  >
    <li
      class="player-list-item list-item"
    >
      <span
        class="columns is-vcentered player-info is-primary is-top is-medium b-tooltip is-animated"
        data-label="foo"
        style="transition-delay: 0ms;"
      >
        <div
          class="column is-narrow"
        >
          <div>
            <img
              src="../assets/logo.png"
            />
          </div>
        </div>
         
        <div
          class="column player-name has-text-left is-size-4"
        >
          <p>
            foo
          </p>
        </div>
         
        <div
          class="column is-narrow"
        >
          <!---->
        </div>
      </span>
    </li>
    <li
      class="player-list-item list-item"
    >
      <span
        class="columns is-vcentered player-info is-primary is-top is-medium b-tooltip is-animated"
        data-label="bar"
        style="transition-delay: 0ms;"
      >
        <div
          class="column is-narrow"
        >
          <div>
            <img
              src="../assets/logo.png"
            />
          </div>
        </div>
         
        <div
          class="column player-name has-text-left is-size-4"
        >
          <p>
            bar
          </p>
        </div>
         
        <div
          class="column is-narrow"
        >
          <!---->
        </div>
      </span>
    </li>
  </ol>
</div>
`;
